### PR TITLE
Embed etcd

### PR DESCRIPTION
**Pull request by prashanthpai**

### CODE REVIEW

**Comment by prashanthpai** [comment]
CI build failure:
~~~
09:33:06 [INFO]	--> Fetching gopkg.in/yaml.v2.
09:33:10 [ERROR]	Failed to install: hg is not installed
09:33:10 make: *** [vendor-update] Error 1
~~~

Install mercurial on test machines:

Fix here:

https://github.com/gluster/glusterfs-patch-acceptance-tests/blob/master/centos-ci/glusterd2/run-test.sh#L22

**Comment by kshlm** [comment]
That is a git repo, don't know why it's failing. I'll try retriggering once more.

File: etcdmgmt/embed.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
+package etcdmgmt
+
+import (
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/coreos/etcd/embed"
```
Comment (kshlm) [inline etcdmgmt/embed.go]: The format I like to follow for imports is to have stdlib packages grouped together, followed by the project packages and then all external packages. Makes it easier to identify packages, and is consistent with the rest of the code.

File: errors/error.go [hunk 2]
```diff
@@ -31,4 +31,6 @@ var (
 	ErrPeerLocalNode           = errors.New("The peer being added is the local node")
 	ErrProcessNotFound         = errors.New("The process is not running or is inaccessible")
 	ErrProcessAlreadyRunning   = errors.New("Process is already running")
+	ErrEtcdStartTimeout        = errors.New("Etcd embedded server took too long to start!")
```
Comment (kshlm) [inline errors/error.go]: This is something I'd like to change. It's better to keep errors within the actual package returning these errors. It doesn't make sense to have a separate errors package (IMO).

File: glide.lock [hunk 3]
```diff
@@ -1,32 +1,119 @@
-hash: 78c31a81cf8d4aa9d05bff8041cfdded835006f3bc3cb75e398f86f015abc716
-updated: 2016-09-23T17:33:15.67392395+05:30
+hash: e4a04ec7ffb626e1ece87c0eed2c144ce3331adcbff4cc2e38984a46ada6c5e0
```
Comment (kshlm) [inline glide.lock]: WTF!! This is a huge number of new dependencies, all due to just one extra package!

File: gdctx/global.go [hunk 4]
```diff
@@ -37,6 +38,7 @@ var (
 	OpVersion      int
 	EtcdProcessCtx *os.Process
 	EtcdClient     etcdclient.Client
+	EtcdInstance   *embed.Etcd
```
Comment (kshlm) [inline gdctx/global.go]: Just keep it `Etcd`, it makes enough sense.

Comment (kshlm) [inline gdctx/global.go]: I'm not sure if we need to keep the etcd instance in `gdctx`. Couldn't we just let it be in the `etcdmgmt` package itself?

**Comment by kshlm** [comment]
> CI build failure:
> 
> 09:33:06 [INFO] --> Fetching gopkg.in/yaml.v2.
> 09:33:10 [ERROR]    Failed to install: hg is not installed
> 09:33:10 make: *** [vendor-update] Error 1
> Install mercurial on test machines:
> 
> Fix here:
> 
> https://github.com/gluster/glusterfs-patch-acceptance-tests/blob/master/centos-ci/glusterd2/run-test.sh#L22

Checked it again. The failure is for a different pacakge `bitbucket.org/ww/goautoneg`, which is a mercurial repo. I'll fix the run_test.sh script.

File: gdctx/global.go (hunk 4 above)
Comment (prashanthpai) [inline gdctx/global.go]: I expect it to be referenced elsewhere in code too (just in my head right now), especially around add and remove peer operations. I'll move it out later if needed.

File: glide.lock (hunk 3 above)
Comment (prashanthpai) [inline glide.lock]: Yeah, I hear you. etcd version changed from `v3.0.8` to `v3.1.0-rc.0`. Most of the diff you see is from new subpackages of etcd.

**Comment by prashanthpai** [comment]
It seems like [embed](https://godoc.org/github.com/coreos/etcd/embed) package does not support embedding an etcd server that is running in proxy mode, yet. Until then, I'll stick to running each  embedded etcd instance as a normal cluster member.

**Comment by prashanthpai** [comment]
This still has an issue with libkv + embedded etcd integration.

**Comment by prashanthpai** [comment]
Restarting embed + client connection issue is tracked here: https://github.com/coreos/etcd/issues/6733

File: glide.yaml [hunk 5]
```diff
@@ -5,7 +5,10 @@ import:
   version: v3.1.0-rc.0
   subpackages:
   - client
-- package: github.com/docker/libkv
+  - clientv3
+  - embed
+- package: github.com/prashanthpai/libkv
```
Comment (prashanthpai) [inline glide.yaml]: This is a stop-gap solution for the etcd client issue.

File: commands/peers/peer-rpc-svc.go [hunk 6]
```diff
@@ -78,173 +74,78 @@ func (p *PeerService) ValidateDelete(nc netctx.Context, args *PeerDeleteReq) (*P
...
-		err := etcdmgmt.CloseEtcdClient()
-		if err != nil {
-			log.WithField("error", err.Error()).Error("Could not stop etcd client.")
-		}
+	// Gracefully shutdown the etcd client
+	err = etcdmgmt.CloseEtcdClient()
+	if err != nil {
+		opRet = -1
+		opError = fmt.Sprintf("Error stopping etcd client.")
+		log.WithField("error", err.Error()).Error("Error stopping etcd client.")
```
Comment (kshlm) [inline commands/peers/peer-rpc-svc.go]: You can use `log.WithError(err)` when you want to add errors to the log.

File: commands/peers/peer-rpc-svc.go [hunk 7]
```diff
@@ -78,173 +74,78 @@ func (p *PeerService) ValidateDelete(nc netctx.Context, args *PeerDeleteReq) (*P
...
-				opRet = -1
-				opError = fmt.Sprintf("Could not able to write etcd configuration")
-				log.WithField("error", err.Error()).Error("Could not able to write etcd configuration")
-				return nil, err
-			}
+			newEtcdConfig.InitialCluster = c.InitialCluster
+			newEtcdConfig.ClusterState = c.ClusterState
+			newEtcdConfig.Name = c.Name
+			newEtcdConfig.Dir = c.Name + ".dir"
+			// TODO: Store config locally for glusterd2 and etcd restarts
```
Comment (kshlm) [inline commands/peers/peer-rpc-svc.go]: I wanted to a release with the embedded etcd, but not having this is a blocker. Shouldn't be too hard do to.

I'll merge this PR, but I'll wait till store/restore of etcd config is done to make the release.

Comment (prashanthpai) [inline commands/peers/peer-rpc-svc.go]: Addressed it in https://github.com/gluster/glusterd2/pull/166

**Pull request merged**